### PR TITLE
Allow empty extras in `pep508-rs` and add more corner case to tests

### LINKED ISSUE #2127: uv pip install does not accept empty extras
@AlexWaygood ran into this in https://github.com/python/typeshed/pull/11517. I didn't see an issue from him, so here one is :-)
```
λ pip install 'pypyp[]'
Requirement already satisfied: pypyp in .../lib/python3.11/site-packages (1.1.0)
λ uv pip install 'pypyp[]'
error: Failed to parse `pypyp[]`
  Caused by: Expected an alphanumeric character starting the extra name, found ']'
pypyp[]
      ^
λ python
Python 3.11.5 (main, Nov  2 2023, 22:21:08) [Clang 15.0.0 (clang-1500.0.40.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import packaging.requirements
>>> packaging.requirements.Requirement("pypyp[]")
<Requirement('pypyp')>
```

### PR DESCRIPTION
**Pull request by SigureMo**
## Summary

Fixes #2127, allow empty extras, and add more corner case to tests

## References

- [PEP 508 grammar](https://peps.python.org/pep-0508/#complete-grammar)

### CODE REVIEW

File: crates/requirements-txt/src/lib.rs [hunk 1]
```diff
@@ -1205,7 +1205,7 @@ mod test {
         }, {
             insta::assert_display_snapshot!(errors, @r###"
             Couldn't parse requirement in `<REQUIREMENTS_TXT>` at position 6
-            Expected an alphanumeric character starting the extra name, found ','
+            Expected either alphanumerical character (starting the extra name) ']' (ending the extras section), found ','
```
Comment (AlexWaygood) [inline crates/requirements-txt/src/lib.rs:1208]:
```suggestion
            Expected either alphanumerical character (starting the extra name) or ']' (ending the extras section), found ','
```

**Review by AlexWaygood — COMMENTED** [review]
A small grammar nit regarding the error message:

**Review by charliermarsh — APPROVED** [review]
Thanks!

File: crates/pep508-rs/src/lib.rs [hunk 2]
```diff
@@ -617,9 +617,48 @@ fn parse_extras(cursor: &mut Cursor) -> Result<Vec<ExtraName>, Pep508Error> {
     let Some(bracket_pos) = cursor.eat_char('[') else {
         return Ok(vec![]);
     };
+    cursor.eat_whitespace();
```
Comment (charliermarsh) [inline crates/pep508-rs/src/lib.rs:620]: Moved this out from the start of the loop, because it's also at the end of the loop.
